dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -234,6 +241,28 @@ public EntryResource(SessionFactory sessionFactory, PermissionsInterface permiss
         topicsApi = new TopicsApi(apiClient);
     }
 
+    @DELETE
+    @Timed
+    @UnitOfWork
+    @Path("/{id}")
+    @Operation(operationId = "deleteEntry", description = "Completely remove an entry from Dockstore.", security = @SecurityRequirement(name = JWT_SECURITY_DEFINITION_NAME))
+    @ApiResponse(responseCode = HttpStatus.SC_OK + "", description = "Successfully deleted the entry", content = @Content(mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = Entry.class)))
+    @ApiResponse(responseCode = HttpStatus.SC_FORBIDDEN + "", description = ENTRY_NOT_DELETABLE_MESSAGE)
+    public Entry deleteEntry(@ApiParam(hidden = true) @Parameter(hidden = true, name = "user")@Auth User user,
+        @ApiParam(value = "Entry to delete.", required = true) @PathParam("id") Long id) {
+        Entry<? extends Entry, ? extends Version> entry = toolDAO.getGenericEntryById(id);
+        checkNotNullEntry(entry);
+        checkCanWrite(user, entry);
+        throwIf(!entry.isDeletable(), ENTRY_NOT_DELETABLE_MESSAGE, HttpStatus.SC_FORBIDDEN);
```

Comment:
Not sure what the answer is, but should this check be moved/copied to `EntryDAO.delete()`?

Looking at `io.dockstore.webservice.resources.UserResource#deleteSelfFromEntries()` when a user is deleted, we delete their entries too, as long as the entries are unpublished (but it's not checking `wasEverPublic`). Maybe that's OK for that edge case (we don't delete many users, if any), but what if new code gets introduced some day that also does a delete and forgets to do this check?

OTOH, at some point you may need code that deletes the entry when you know what you're doing.

I'm more raising questions here than definitely opining.

Reply:
> Not sure what the answer is, but should this check be moved/copied to `EntryDAO.delete()`?
> 
> Looking at `io.dockstore.webservice.resources.UserResource#deleteSelfFromEntries()` when a user is deleted, we delete their entries too, as long as the entries are unpublished (but it's not checking `wasEverPublic`). Maybe that's OK for that edge case (we don't delete many users, if any), but what if new code gets introduced some day that also does a delete and forgets to do this check?
> 
> OTOH, at some point you may need code that deletes the entry when you know what you're doing.
> 
> I'm more raising questions here than definitely opining.

My take on the DAO methods is that they are for raw, low-level inspection and manipulation.  As you mention, we may need to make an Entry go away for whatever reason, irrespective of its state, and we'd use the DAO `delete` make it happen.  So, at least one DAO `delete` should remain "unconditional".

I considered making a helper to do the entry+event removal, we're near the cusp of needing one, definitely if we add another delete somewhere else.  A [backup] `isDeletable` check would be appropriate there.

Might be good to convert the user deletion code to verify `isDeletable` on the user's entries, if we're ok with the more stringent requirements.